problems/cipher.js

```diff
@@ -0,0 +1,259 @@
+ * We use % to convert the index to its corresponding index when wrapped around the alphabet.
+ * For instance, an index of 40 % 26 = 14, so we use the new index of 14 to find the correct
+ * character.
+ * @param  {int} index    integer representing the index of the character
+ * @param  {str} alphabet an alphabet string to match the index of the character against
+ * @return {str}          a single character at that position in the alphabet
+ */
+const _indexToChar = (index, alphabet) => {
+    // If we're given a negative index, we need to add the alphabet length
+    // to convert it to its positive match
+    if (index < 0) {
+        index = index + alphabet.length;
+    }
+    // We need to use % to make sure the index will fit on the alphabet
+    // e.g., if we're given an index of 40, that doesn't fit on the alphabet.
+    const convertedIndex = index % alphabet.length;
+
+    return alphabet.charAt(convertedIndex);
+};
+
+/**
+ * Validates a given key to ensure that the key is long enough to be able to
+ * properly encode or decode the stringToConvert. If the key is not long enough,
+ * this method will repeat the key characters until the key becomes long enough.
+ * For instance, given a key of 'ab' and a stringToConvert of 'sahar', the key will
+ * return as 'ababa'.
+ * // TO DO: Add more validation for non-alpha keys
+ * @param  {str} key            the key provided by the user
+ * @param  {str} stringToConvert the string we're taking action on to convert later
+ * @return {str}                the validated key
+ */
+const _validateKey = (key, stringToConvert) => {
+    if (key.length != stringToConvert.length) {
+        index = -1;
+        while (key.length != stringToConvert.length) {
+            index += 1;
+            key = key.concat(key.charAt(index));
+        }
+    }
+    return key;
+};
+
+/**
+ * The opposite of decode.
+ * Takes in a string and returns an encoded string. Uses the provided key
+ * to encode each individual character according to the matching key character
+ * (by index). For instance, given a key of 'bbbb': string = 'abcd' returns 'bcde'
+ * @param  {str} stringToEncode user inputted string to encode
+ * @return {str}                encoded string
+ */
+Cipher.prototype.encode = function(strToEncode) {
+    let encodedArray = [];
+    let charStrIndex = -1;
+
+    let key = _validateKey(this.key, strToEncode);
+
+    strToEncode.split('').forEach((char) => {
+        charStrIndex += 1;
+        // get the alphaIndex of the current character
+        const charAlphaIndex = _charToAlphaIndex(char, this.alphabet);
+        // get the alphaIndex of the key character at the current character's position
+        const keyCharAlphaIndex = _charToAlphaIndex(key.toLowerCase().charAt(charStrIndex), this.alphabet);
+
+        const encodedIndex = charAlphaIndex + keyCharAlphaIndex;
+        const encodedChar = _indexToChar(encodedIndex, this.alphabet);
+
+        return encodedArray.push(encodedChar);
```

Comment:
if we use `[].map()` instead of `[].forEach()`, we won't need to do this push and no need to create the encoded array.

Reply:
map returns a new array: 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map